Adding Dynamic Conv and ConvT ops/modules

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68282
* __->__ #68176

Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

HDCharles committed:
WIP Dynamic Conv

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "WIP Dynamic Conv"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,178 @@
+  // Quantize input
+  at::Tensor q_input = at::quantize_per_tensor(
+      input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
```
This code can be factored out in utils that can be shared between linear_dynamic and this one.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,178 @@
+    bool reduce_range) {
+  // On empty input, no output data will be generated,
+  // so use arbitrary qparams.
+  float x_min = 0;
+  float x_max = 0;
+  // Otherwise...
+  if (input.numel() > 0) {
+    x_min = input.min().item<float>();
+    x_max = input.max().item<float>();
+  }
+
+  // Input tensor is quantized as 8-bit unsigned values
+  static constexpr int precision = 8;
+  static constexpr bool is_signed = false;
+
+  // Calculate scale and zero point for quantization of input tensor
+  auto q_params = quant_utils::ChooseQuantizationParams(
+      /*min=*/x_min,
+      /*max=*/x_max,
+      /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+      /*qmax=*/
+      is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+      /*preserve_sparsity=*/false,
+      /*force_scale_power_of_two=*/false,
+      /*reduce_range=*/reduce_range);
+
+  // Quantize input
+  at::Tensor q_input = at::quantize_per_tensor(
+      input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
```
Same here. Both of these look exactly the same. Can you file a task and assign it to me? We dont have to fix this as part of this PR but this needs to be tracked.

HDCharles replied:
ok, will do

HDCharles replied:
https://www.internalfb.com/tasks?p_nav=SUBSCRIBED&t=105737663

kimishpatel commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -889,6 +889,187 @@ def test_embedding_bag_api(self, num_embeddings, embedding_dim, num_offsets, set
+        np.testing.assert_array_almost_equal(
+            Y.numpy(), Y_loaded.numpy(), decimal=0)
+
+        # Test serialization
+        b = io.BytesIO()
+        torch.save(dynamic_module, b)
+        b.seek(0)
+        loaded_conv = torch.load(b)
+
+        self.assertEqual(loaded_conv.bias(), dynamic_module.bias())
+        self.assertEqual(loaded_conv.scale, dynamic_module.scale)
+        self.assertEqual(loaded_conv.zero_point,
+                         dynamic_module.zero_point)
```
Didnt the lines above already test serialization?

HDCharles replied:
I'm just copying from the qlinear_dynamic test, not 100% sure what the specific corner case its looking to check

HDCharles replied:
ok I can just fix

kimishpatel commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -889,6 +889,187 @@ def test_embedding_bag_api(self, num_embeddings, embedding_dim, num_offsets, set
+        self.assertEqual(loaded_conv.zero_point,
+                         dynamic_module.zero_point)
+
+        # Test copy and deepcopy
+        copied_conv = copy.copy(dynamic_module)
+        self.assertEqual(copied_conv.bias(), dynamic_module.bias())
+        self.assertEqual(copied_conv.scale, dynamic_module.scale)
+        self.assertEqual(copied_conv.zero_point,
+                         dynamic_module.zero_point)
+        Y_copied = copied_conv(X_fp32, reduce_range)
+        np.testing.assert_array_almost_equal(
+            Y.numpy(), Y_copied.numpy(), decimal=0)
+
+        deepcopied_conv = copy.deepcopy(dynamic_module)
+        self.assertEqual(deepcopied_conv.bias(), dynamic_module.bias())
+        self.assertEqual(deepcopied_conv.scale, dynamic_module.scale)
+        self.assertEqual(deepcopied_conv.zero_point,
+                         dynamic_module.zero_point)
+        Y_deepcopied = copied_conv(X_fp32, reduce_range)
+        np.testing.assert_array_almost_equal(
+            Y.numpy(), Y_deepcopied.numpy(), decimal=0)
```
Do you know why these tests are needed?

HDCharles replied:
my understanding is taht in general quantized modules often have issues with this. However I'm just copying the test_linear_api from this same test class

kimishpatel commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -889,6 +889,187 @@ def test_embedding_bag_api(self, num_embeddings, embedding_dim, num_offsets, set
+        conv_module.qconfig = torch.ao.quantization.default_dynamic_qconfig  # type: ignore[assignment]
+        prepare_dynamic(conv_module)
+        conv_module(X_dq)
+        quantized_conv_module = dq_mod.from_float(conv_module)
```
Is this dynamically quantized conv module? If so change names including conv_module one.

Also do you need to check for functional correctness?

HDCharles replied:
no its the unquantized conv module, this is testing the quantization process. dynamic_module, quantized_module and conv_module are the dynamically quantized, statically quantized, and fp32 modules respectively.

kimishpatel commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3661,6 +3661,93 @@ def test_embedding_bag_2d_indices(self):
+        if q_engine not in torch.backends.quantized.supported_engines:
+            return
+        torch.backends.quantized.engine = q_engine
+        reduce_range = True
```
This should be true only for fbgemm.

HDCharles replied:
I fixed this to work correctly and discard reduce_range in qnnpack

kimishpatel commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3661,6 +3661,93 @@ def test_embedding_bag_2d_indices(self):
 
 
+class TestDynamicQuantizedConv(TestCase):
+    def _test_qconv_op_impl(self, q_mod, dq_op, dim, q_engine, dtype):
```
Nice!

kimishpatel commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3661,6 +3661,93 @@ def test_embedding_bag_2d_indices(self):
+        dtype = torch.quint8
+
+        for q_engine in ["fbgemm", "qnnpack"]:
+            for i in range(10):
```
Why run 10 times? Just with different rand inputs? If so I dont suppose we need to do that.

HDCharles replied:
ok, removed

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,370 @@
+    def _get_name(self):
+        return 'DynamicQuantizedConv1d'
+
+    # @reviewer unsure whether reduce range should be true or false, here I mimic qlinear_dynamic using true
```
This should be true for fbgemm. Isnt that how qlinear_dynamic also handles this, https://fburl.com/mc19gzcn. I think supplying this is fine since qnnpack will ignore this anyway. But API is definitely not good. One would expect reduce range to be honored if supplied. Not ignored depending on the backend.

HDCharles replied:
I further aligned it to qlinear_dynamic, especially in c++ this implementation was different, now it discards reduce_range for qnnpack

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,370 @@
+        return 'DynamicQuantizedConv2d'
+
+    # @reviewer unsure whether reduce range should be true or false, here I mimic qlinear_dynamic using true
+    def forward(self, input: Tensor, reduce_range: bool = True) -> Tensor:
```
Also reduce_range should really be the property of the module so that depending on fbgemm qconfig or qnnpack it can be set appropriately. cc: @vkuzo

HDCharles replied:
Yeah I asked Vasiliy about this and he said it would make sense to just do what dynamic quantized linear does and at some point we should update all the APIs at the same time, will file a github issue

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,370 @@
+            False, _triple(0), groups, bias, padding_mode, **factory_kwargs)
+
+    def _get_name(self):
+        return 'QuantizedConv3d'
```
Dynamic?

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,370 @@
+            groups, bias, dilation, padding_mode, **factory_kwargs)
+
+    def _get_name(self):
+        return 'QuantizedConvTranpose2d'
```
Dynamic?

kimishpatel reviewed (changes requested):
Looks good for the most part. Left some minor comments.

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3661,6 +3661,93 @@ def test_embedding_bag_2d_indices(self):
         torch.testing.assert_close(result, qresult, atol=0.05, rtol=1e-3)
 
 
+class TestDynamicQuantizedConv(TestCase):
```
nit can we put this to TestQuantizedOps or TestDynamicQuantizedOps? we'd like to reduce the TestCases in this file actually

HDCharles replied:
ok, will do

HDCharles replied:
actually I'm going to leave as is, makes more sense to factor that out in a different PR

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,178 @@
+  TORCH_CHECK(
+      fbgemm::fbgemmSupportedCPU(), "Your CPU does not support FBGEMM.");
+
+  float x_min, x_max;
+  fbgemm::FindMinMax(
+      /*m=*/input.data_ptr<float>(),
+      /*min=*/&x_min,
+      /*max=*/&x_max,
+      /*len=*/input.numel());
+
+  // Input tensor is quantized as 8-bit unsigned values
+  static constexpr int precision = 8;
+  static constexpr bool is_signed = false;
+
+  // Calculate scale and zero point for quantization of input tensor
+  auto q_params = quant_utils::ChooseQuantizationParams(
+      /*min=*/x_min,
+      /*max=*/x_max,
+      /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+      /*qmax=*/
+      is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+      /*preserve_sparsity=*/false,
+      /*force_scale_power_of_two=*/false,
+      /*reduce_range=*/reduce_range);
+
+  // Quantize input
+  at::Tensor q_input = at::quantize_per_tensor(
+      input, q_params.scale, q_params.zero_point, c10::kQUInt8);
```
I believe there is an operator now that should be able to do this, i.e `quantize_per_tensor_dynamic`. cc @jerryzh168 and @dzdang  can you confirm?

jerryzh168 replied:
oh right, it's landing currently: https://github.com/pytorch/pytorch/pull/68004/files

Looks like it's not using fbgemm::FindMinMax currently, maybe we can speed it up with that

kimishpatel replied:
I think Charles filed a task for this.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,178 @@
+  // transpose
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::conv_transpose1d_dynamic"),
+      TORCH_FN(QConv1dDynamicInt8::run));
```
why is conv transpose using the exact same class as conv?

kimishpatel replied:
I dont remember the details and @z-a-f added it probably. But for the most part it is likely because they share much of the functionality.

HDCharles replied:
yes, its the same as for the statically quantized ops

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -232,7 +232,8 @@ def from_float(cls, mod):
             assert hasattr(mod, "qconfig"), \
                 "Input float module must have qconfig defined."
-            activation_post_process = mod.activation_post_process
+            activation_post_process = None if not hasattr(
```
in get_qconv L206 in this file, can we check if activation_post_process is None or not, instead of using try/exception?
and we can also skip assigning scale and zero_point when activation_post_process is None as well

HDCharles replied:
no because the implementation has activation_post_process but it just raises an exception, will look at not assigning scale/zp

HDCharles replied:
added a scale/zp skip

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

kimishpatel reviewed (approved):
LGTM

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -205,13 +205,11 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
         qconv.set_weight_bias(qweight, mod.bias)
         try:
             act_scale, act_zp = activation_post_process.calculate_qparams()
+            qconv.scale = float(act_scale)
```
Sorry, why can't we remove the try/except here? can we do something like
```
if activation_post_process is not None:
    ...
else:
    ...
```

HDCharles replied:
because it does have a activation_post_process, it just raises an error when you try to calculate_qparams

https://github.com/pytorch/pytorch/blob/f455030931c30bfeaab7f79b3f4305000674d6f2/torch/ao/quantization/observer.py#L1215

vkuzo replied:
I think if we want a try-except here, the placeholder observer could throw a more specific exception type and this except clause would catch just that type.

I'd vote +1 on removing the try-catch cause, it feels cleaner to never insert an incorrect observer in the first place.  We should make logic to check that observers are correct in the prepare pass, not in the convert pass.

vkuzo replied:
There is some logic here (https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/qconfig.py#L247) for checking whether a qconfig is valid for a module, perhaps this could be extended to cover this case.

HDCharles replied:
I edited it to check for activation post process dtype, which feels like it should be sufficient.

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles committed:
Update on "Adding Dynamic Conv and ConvT ops/modules"


Summary: it should be noted that for the modules, reduce_range is set to
true by default in a similar fashion to linear_dynamic.

other notes:

the primary changes occur in qconv_dynamic.cpp and nn/quantized/dynamic/modules/conv.py, in qconv_dynamic.cpp there is both an FBGEMM and a QNNPACK implementation which simply calculates the q_params from the input and then calls the relevant quantized conv function and dequantizing the output. The modules in ../dynamic/modules/conv.py inherit from the static modules in nn/quantized/modules/conv.py and since dynamic modules require fewer parameters than statically quantized modules, the only things that they aren't able to share are the forward functions. c++ ops are registered in library.cpp

new tests are added to make sure the implementation is numerically accurate and also stable. There tests mimic the tests for both dynamic_linear tests and quantized_conv.

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule
python test/test_quantization.py TestDynamicQuantizedConv
python test/test_quantization.py TestQuantizedConv

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32374003](https://our.internmc.facebook.com/intern/diff/D32374003)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32374003).
